<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [7]:
import pandas as pd
import psycopg2

In [8]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [9]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [10]:
# текст запроса
query_3_1 = f'''
select count(id)
from public.vacancies
'''

In [11]:
# результат запроса
count_vacancies = pd.read_sql_query(query_3_1, connection)
display (count_vacancies)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [12]:
# текст запроса
query_3_2 = f'''
select count(id)
from public.employers
'''

In [13]:
# результат запроса
count_employers = pd.read_sql_query(query_3_2, connection)
display (count_employers)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [14]:
# текст запроса
query_3_3 = f'''
select count(id)
from public.areas
'''


In [15]:
# результат запроса
count_areas = pd.read_sql_query(query_3_3, connection)
display (count_areas)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [16]:
# текст запроса
query_3_4 = f'''
select count(id)
from public.industries
'''

In [17]:
# результат запроса
count_industries = pd.read_sql_query(query_3_4, connection)
display (count_industries)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

***

### выводы по предварительному анализу данных

В нашей базе данных 49197 уникальных вакансий  от  23501 работадателей из 1362 регионов . Работадатели представляют 294 сферы деятельности  

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [19]:
# текст запроса
query_4_1 = f'''
select count(v.id) cnt,
a.name area
from public.vacancies v
join public.areas a on v.area_id = a.id
group by area
order by cnt desc
limit 5
'''

In [20]:
# результат запроса
top_areas = pd.read_sql_query(query_4_1, connection)
display (top_areas)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [21]:
# текст запроса
query_4_2 = f'''
select count(id) cnt
from public.vacancies v
where v.salary_from is not Null or v.salary_to is not Null
'''

In [22]:
# результат запроса
vac_salary = pd.read_sql_query(query_4_2, connection)
display (vac_salary)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [23]:
# текст запроса
query_4_3 = f'''
select avg(salary_from) salary_from,
avg(salary_to) salary_to
from public.vacancies v
'''

In [24]:
# результат запроса
df_sal = pd.read_sql_query(query_4_3, connection)
display (df_sal)
#display(df_sal['salary_from'].mean())
#display(df_sal['salary_to'].mean())

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


salary_from      salary_to
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [25]:
# текст запроса
query_4_4 = f'''
select schedule,
employment,
count (v.id) cnt
from public.vacancies v
group by schedule, employment
order by cnt desc
limit 3
'''

In [26]:
# результат запроса
df_sh_em = pd.read_sql_query(query_4_4, connection)
display (df_sh_em)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule        employment    cnt
0       Полный день  Полная занятость  35367
1  Удаленная работа  Полная занятость   7802
2     Гибкий график  Полная занятость   1593

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [27]:
# текст запроса
query_4_5 = f'''
select experience,
count (v.id) cnt
from public.vacancies v
group by experience
order by cnt
'''

In [28]:
# результат запроса
df_exp = pd.read_sql_query(query_4_5, connection)
display (df_exp)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [29]:
# выводы по детальному анализу вакансий

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [30]:
# текст запроса
query_5_1 = f'''
select e.name,
count (v.id) cnt
from public.vacancies v
join public.employers e on v.employer_id = e.id
group by e.name
order by cnt desc
limit 6
'''

In [31]:
# результат запроса
df_emp = pd.read_sql_query(query_5_1, connection)
display (df_emp)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331
5    ИК СИБИНТЕК   327

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [32]:
# текст запроса
query_5_2 = f'''
select a.name,
count (v.id) v_cnt,
count (e.id) e_cnt
from public.areas a
left join public.vacancies v on v.area_id = a.id
left join public.employers e on a.id = e.area
group by a.name
having count(v.id) = 0
order by e_cnt desc
'''

In [33]:
# результат запроса
df_area = pd.read_sql_query(query_5_2, connection)
display (df_area)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  v_cnt  e_cnt
0                          Россия      0    410
1                       Казахстан      0    207
2              Московская область      0     75
3              Краснодарский край      0     19
4              Ростовская область      0     18
..                            ...    ...    ...
588                      Струнино      0      0
589    Славгород (Алтайский край)      0      0
590                     Сорочинск      0      0
591  Ленинский (Тульская область)      0      0
592                         Тавда      0      0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [34]:
# текст запроса
query_5_3 = f'''
select e.name,
count (distinct v.area_id) a_cnt

from public.employers e
left join public.vacancies v on v.employer_id = e.id
group by e.name
order by a_cnt desc
'''


In [35]:
# результат запроса
df_emp_area = pd.read_sql_query(query_5_3, connection)
display (df_emp_area)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  a_cnt
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
23170                  СДЕЛКА      0
23171            Alandr Group      0
23172                     СДК      0
23173      Lemon Land Lombard      0
23174                  Махина      0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [36]:
# текст запроса
query_5_4 = f'''
select count( distinct e.id)
from  public.employers e
left join public.employers_industries ei on e.id = ei.employer_id 
where industry_id is null
'''

In [37]:
# результат запроса
df_null_ind = pd.read_sql_query(query_5_4, connection)
display (df_null_ind)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [38]:
# текст запроса
query_5_5 = f'''
select e.name,
count(*)
from  public.employers e
left join public.employers_industries ei on e.id = ei.employer_id
group by  e.name
having count( distinct industry_id) = 4
order by name
limit 1
offset 2
'''

In [39]:
# результат запроса
df_4ind = pd.read_sql_query(query_5_5, connection)
display (df_4ind)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [40]:
# текст запроса
query_5_6 = f'''
select count( distinct e.id)
from  public.employers e
left join public.employers_industries ei on e.id = ei.employer_id
join industries i on i.id=ei.industry_id
where i.name ='Разработка программного обеспечения'
'''

In [41]:
# результат запроса
df_ind_name = pd.read_sql_query(query_5_6, connection)
display (df_ind_name)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [42]:
# код для получения списка городов-милионников

import urllib.parse as req
import requests

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

cities = list(pd.read_html(url)[1]['Город'])
cities[-3] = 'Воронеж'
cities = tuple(cities)
cities


('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [43]:
# текст запроса
query_5_7 = f'''
select a.name,
count(v.id) cnt
from public.areas a
join public.vacancies v on v.area_id = a.id
join public.employers e on v.employer_id = e.id
where e.name = 'Яндекс'
group by a.name
having a.name in {cities}

union all
select 
'Total',
sum(b.cnt)
from 
(
select a.name,
count(v.id) cnt
from public.areas a
join public.vacancies v on v.area_id = a.id
join public.employers e on v.employer_id = e.id
where e.name = 'Яндекс'
group by a.name
having a.name in {cities}
) as b
order by cnt
'''

In [44]:
# результат запроса
df_ya = pd.read_sql_query(query_5_7, connection)
display (df_ya)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7            Самара   26.0
8               Уфа   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

In [45]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [46]:
# текст запроса
query_6_1 = f'''
select count(v.name)
from  public.vacancies v

where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
'''

In [47]:

# результат запроса
df_data = pd.read_sql_query(query_6_1, connection)
display (df_data)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [62]:
# текст запроса
query_6_2 = f'''
select count(v.id)

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and
(
lower(v.name) like '%junior%' or
v.experience like '%Нет опыта%' or
v.employment like '%Стажировка%'
)

'''

In [63]:
# результат запроса
df_ds = pd.read_sql_query(query_6_2, connection)
display (df_ds)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [50]:
# текст запроса
query_6_3 = f'''
select v.id

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
v.name like '%ML%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and
(
lower(v.key_skills) like '%sql%' or
lower(v.key_skills) like '%postgres%'
)
except
select v.id

from
public.vacancies v

where
v.name like '%HTML%'
'''

In [51]:
# результат запроса
df_sql = pd.read_sql_query(query_6_3, connection)
display (df_sql)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


id
0    54633834
1    54559086
2    54827065
3    55120846
4    54592046
..        ...
196  55135290
197  55127397
198  54938477
199  54974571
200  55135252

[201 rows x 1 columns]

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [64]:
# текст запроса
query_6_4 = f'''
select count(v.id)

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and
(
lower(v.key_skills) like '%python%' 
)

'''

In [65]:
# результат запроса
df_py = pd.read_sql_query(query_6_4, connection)
display (df_py)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [55]:
query_6_5 = f'''
select 
avg (char_length( v.key_skills)- char_length(replace (v.key_skills, CHR(9),'' ))+1 )   avg

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
) 
and v.key_skills is not null

'''

In [56]:
# результат запроса
df_ds_skill = pd.read_sql_query(query_6_5, connection)
display (df_ds_skill)

                    


c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avg
0  6.406032

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [57]:
query_6_6 = f'''
select 
b.experience ,
avg(b.sallary)
from 
(
select 
v.experience experience,
coalesce(v.salary_from, 0) + coalesce(v.salary_to, 0) sallary
from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
) 

and 

(
    (v.salary_from is null and v.salary_to is not null)
    or (v.salary_from is not null and v.salary_to is  null)
)


union all

select 
v.experience,
(v.salary_from + v.salary_to)/2
from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
) 

and 

(
    (v.salary_from is not null and v.salary_to is not null)
 
)


) as b

group by 
b.experience

'''

In [58]:
# результат запроса
df_ds_avg = pd.read_sql_query(query_6_6, connection)
display (df_ds_avg)


c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience            avg
0           Нет опыта   74642.857143
1  От 1 года до 3 лет  139674.750000
2       От 3 до 6 лет  243114.666667

***

In [59]:
# выводы по предметному анализу

# Общий вывод по проекту

In [60]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования